lib/src/main/kotlin/at/bitfire/ical4android/util/AndroidTimeUtils.kt

```diff
@@ -142,104 +147,141 @@ object AndroidTimeUtils {
             event with time     (undefined)                         store as ...ThhmmssZ
         */
         val dateFormatUtcMidnight = SimpleDateFormat("yyyyMMdd'T'000000'Z'", Locale.ROOT)
-        val strDates = LinkedList<String>()
 
-        // use time zone of first entry for the whole set; null for UTC
-        val tz =
-            (dates.firstOrNull() as? RDate)?.periods?.timeZone ?:   // VALUE=PERIOD (only RDate)
-            dates.firstOrNull()?.dates?.timeZone                    // VALUE=DATE/DATE-TIME
+        /** Provides a collection of dates for each timezone id (key) */
+        val datesMap = mutableMapOf<String, MutableList<String>>()
 
         for (dateListProp in dates) {
             if (dateListProp is RDate && dateListProp.periods.isNotEmpty()) {
                 Ical4Android.log.warning("RDATE PERIOD not supported, ignoring")
                 break
             }
 
-            when (dateListProp.dates.type) {
+            // Tries to get the timezone from the property
+            val tz: TimeZone? =
+                (dateListProp as? RDate)?.periods?.timeZone ?:   // VALUE=PERIOD (only RDate)
+                dateListProp.dates?.timeZone                     // VALUE=DATE/DATE-TIME
+            // UTC is used by default if no timezone has been set
+            val tzId = tz?.id ?: TimeZones.UTC_ID
+
+            // Extract a list of dates from the property
+            val datesList = when (dateListProp.dates.type) {
                 Value.DATE_TIME -> {
                     if (tz == null && !dateListProp.dates.isUtc)
                         dateListProp.setUtc(true)
                     else if (tz != null && dateListProp.timeZone != tz)
                         dateListProp.timeZone = tz
 
                     if (allDay)
-                        dateListProp.dates.mapTo(strDates) { dateFormatUtcMidnight.format(it) }
+                        dateListProp.dates.map { dateFormatUtcMidnight.format(it) }
                     else
-                        strDates.add(dateListProp.value)
+                        listOf(dateListProp.value)
```

Comment:
This whole section is a bit obscure. I don't understand anymore when tz is null, what dateListProp.timeZone is etc… I think we have to do it more verbosely and properly. Chances are high that it will break something.

Also, `dateListProp.value` shouldn't be used, but the date-times should be mapped and formatted like for all-day.